### PR TITLE
WalletConnectUnity integration

### PR DESCRIPTION
**Pull request by oleksandrchainsafe**
WIP

### CODE REVIEW

File: Packages/io.chainsafe.web3-unity/Runtime/WalletConnect/WalletConnectUnityProvider.cs [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+        public override Task Disconnect()
+        {
+            WalletConnectModal.Disconnect();
+            return Task.CompletedTask;
+        }
+
+        public override Task<T> Perform<T>(string method, params object[] parameters)
+        {
+            throw new NotImplementedException();
```
Comment (oleksandrchainsafe) [inline Packages/io.chainsafe.web3-unity/Runtime/WalletConnect/WalletConnectUnityProvider.cs:121]: This needs to be implemented.

File: Packages/io.chainsafe.web3-unity.walletconnect-chainsafe/package.json [hunk 2]
```diff
@@ -0,0 +1,25 @@
+{
+  "name": "io.chainsafe.web3-unity.walletconnect-chainsafe",
```
Comment (kantagara) [inline Packages/io.chainsafe.web3-unity.walletconnect-chainsafe/package.json:2]: wait didn't we say that we're leaving it as part of our core package

File: Packages/io.chainsafe.web3-unity.walletconnect/package.json [hunk 3]
```diff
@@ -0,0 +1,25 @@
+{
+  "name": "io.chainsafe.web3-unity.walletconnect",
+  "version": "1.0.0",
+  "displayName": "web3.unity SDK WalletConnect",
+  "description": "web3.unity is an open-source gaming SDK written in C# and developed by ChainSafe Gaming. It connects games built in the Unity game engine to the blockchain. The library currently supports games built for web browsers (WebGL), iOS/Android mobile, and desktop. web3.unity is compatible with most EVM-based chains such as Ethereum, Polygon, Moonbeam, Cronos, Nervos, and Binance Smart Chain, letting developers easily choose and switch between them to create the best in-game experience.",
+  "license": "LGPL-3.0-only",
+  "licensesUrl": "https://github.com/ChainSafe/web3.unity/blob/main/LICENSE",
+  "documentationUrl": "https://docs.gaming.chainsafe.io/",
+  "dependencies": {
+    "io.chainsafe.web3-unity": "2.5.5",
```
Comment (kantagara) [inline Packages/io.chainsafe.web3-unity.walletconnect/package.json:10]: 2.6

File: Packages/io.chainsafe.web3-unity/package.json [hunk 4]
```diff
@@ -1,32 +1,33 @@
...
+  "version": "2.5.5",
+  "displayName": "web3.unity SDK",
+  "description": "web3.unity is an open-source gaming SDK written in C# and developed by ChainSafe Gaming. It connects games built in the Unity game engine to the blockchain. The library currently supports games built for web browsers (WebGL), iOS/Android mobile, and desktop. web3.unity is compatible with most EVM-based chains such as Ethereum, Polygon, Moonbeam, Cronos, Nervos, and Binance Smart Chain, letting developers easily choose and switch between them to create the best in-game experience.",
+  "license": "LGPL-3.0-only",
+  "licensesUrl": "https://github.com/ChainSafe/web3.unity/blob/main/LICENSE",
+  "documentationUrl": "https://docs.gaming.chainsafe.io/",
+  "dependencies": {
+    "com.unity.nuget.newtonsoft-json": "3.0.2",
+    "com.unity.textmeshpro": "3.0.6",
+    "com.walletconnect.modal": "1.1.6"
```
Comment (kantagara) [inline Packages/io.chainsafe.web3-unity/package.json:12]: aha so this will get the dependency itself?

**Review by kantagara — APPROVED** [review]
code looks good. Haven't tested it yet

**Review by juans-chainsafe — CHANGES_REQUESTED** [review]
I have this build error:

`/Users/juanmanuelspoleti/Desktop/workspace/web3.unity/Packages/io.chainsafe.web3-unity/Runtime/WalletConnect/WalletConnectUnityConfigAsset.cs(7,68): error CS0535: 'WalletConnectUnityConfigAsset' does not implement interface member 'IWalletConnectUnityConfig.ProjectId'`

**Review by juans-chainsafe — CHANGES_REQUESTED** [review]
After you click play in Unity Editor, this is the error:

```
MissingFieldException: Field not found: string WalletConnectSharp.Core.Models.CoreOptions.Name Due to: Could not find field in class
System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[TStateMachine] (TStateMachine& stateMachine) (at <e307bbb467104258887a104f6151f183>:0)
ChainSafe.Gaming.WalletConnect.WalletConnectProvider.ChainSafe.Gaming.Web3.Core.ILifecycleParticipant.WillStartAsync () (at <1e67fc4a4e504efead5c3fa462bcc325>:0)
ChainSafe.Gaming.Web3.Web3.InitializeAsync () (at <b140efe57f1547d1b01d0a207aa98c36>:0)
ChainSafe.Gaming.Web3.Build.Web3Builder.LaunchAsync () (at <b140efe57f1547d1b01d0a207aa98c36>:0)
WalletConnectLoginProvider.Initialize () (at Assets/Samples/web3.unity SDK WalletConnect (ChainSafe)/1.0.0/web3.unity SDK WalletConnect (ChainSafe) Samples/WalletConnectLoginProvider.cs:25)
System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__7_0 (System.Object state) (at <e307bbb467104258887a104f6151f183>:0)
UnityEngine.UnitySynchronizationContext+WorkRequest.Invoke () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:153)
UnityEngine.UnitySynchronizationContext.Exec () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:83)
UnityEngine.UnitySynchronizationContext.ExecuteTasks () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:107)
```

Also, we should remove the old WalletConnect scene from the Scenes in Build section
Screenshot 2024-05-30 at 4 00 50 PM

File: Packages/io.chainsafe.web3-unity.walletconnect-chainsafe/package.json (hunk 2 above)
Comment (oleksandrchainsafe) [reply Packages/io.chainsafe.web3-unity.walletconnect-chainsafe/package.json:2]: we've decided we gonna keep the new one in the core and provide the one created by me in a separate package.

File: Packages/io.chainsafe.web3-unity/package.json (hunk 4 above)
Comment (oleksandrchainsafe) [reply Packages/io.chainsafe.web3-unity/package.json:12]: yeah, this references the UPM package for their solution

**Comment by oleksandrchainsafe** [comment]
It is planned to be released with 2.6.1

**Comment by oleksandrchainsafe** [comment]
Fixed branch last week. Juan is testing the success rate for the connection of different wallets.

**Comment by juans-chainsafe** [comment]
> Fixed branch last week. Juan is testing the success rate for the connection of different wallets.

On it! In the mean time you can fix the recent merge conflicts that appear

**Review by juans-chainsafe — CHANGES_REQUESTED** [review]
Build errors are fixed, but Im not able to login with Wallet connect, not even with Metamask. I scan the QR code of Metamask, and I get these errors:

```
[Sample Test-engine] response was error
UnityEngine.Debug:LogError (object)
WalletConnectUnity.Core.Logger:LogError (string) (at ./Library/PackageCache/com.walletconnect.core@3.1.6/Runtime/Logger.cs:11)
WalletConnectSharp.Common.Logging.WrapperLogger:LogError (string)
WalletConnectSharp.Sign.Engine/<WalletConnectSharp-Sign-Interfaces-IEnginePrivate-OnSessionProposeResponse>d__108:MoveNext ()
System.Runtime.CompilerServices.AsyncTaskMethodBuilder:Start<WalletConnectSharp.Sign.Engine/<WalletConnectSharp-Sign-Interfaces-IEnginePrivate-OnSessionProposeResponse>d__108> (WalletConnectSharp.Sign.Engine/<WalletConnectSharp-Sign-Interfaces-IEnginePrivate-OnSessionProposeResponse>d__108&)
WalletConnectSharp.Sign.Engine:WalletConnectSharp.Sign.Interfaces.IEnginePrivate.OnSessionProposeResponse (string,WalletConnectSharp.Network.Models.JsonRpcResponse`1<WalletConnectSharp.Sign.Models.Engine.Methods.SessionProposeResponse>)
WalletConnectSharp.Core.Controllers.TypedMessageHandler/<>c__DisplayClass18_0`2/<<HandleMessageType>g__ResponseCallback|1>d<WalletConnectSharp.Sign.Models.Engine.Methods.SessionPropose, WalletConnectSharp.Sign.Models.Engine.Methods.SessionProposeResponse>:MoveNext ()
System.Runtime.CompilerServices.AsyncVoidMethodBuilder:Start<WalletConnectSharp.Core.Controllers.TypedMessageHandler/<>c__DisplayClass18_0`2/<<HandleMessageType>g__ResponseCallback|1>d<WalletConnectSharp.Sign.Models.Engine.Methods.SessionPropose, WalletConnectSharp.Sign.Models.Engine.Methods.SessionProposeResponse>> (WalletConnectSharp.Core.Controllers.TypedMessageHandler/<>c__DisplayClass18_0`2/<<HandleMessageType>g__ResponseCallback|1>d<WalletConnectSharp.Sign.Models.Engine.Methods.SessionPropose, WalletConnectSharp.Sign.Models.Engine.Methods.SessionProposeResponse>&)
WalletConnectSharp.Core.Controllers.TypedMessageHandler/<>c__DisplayClass18_0`2<WalletConnectSharp.Sign.Models.Engine.Methods.SessionPropose, WalletConnectSharp.Sign.Models.Engine.Methods.SessionProposeResponse>:<HandleMessageType>g__ResponseCallback|1 (object,WalletConnectSharp.Core.Models.Relay.MessageEvent)
WalletConnectSharp.Core.Controllers.TypedMessageHandler/<>c__DisplayClass18_0`2/<<HandleMessageType>g__InspectResponseRaw|2>d<WalletConnectSharp.Sign.Models.Engine.Methods.SessionPropose, WalletConnectSharp.Sign.Models.Engine.Methods.SessionProposeResponse>:MoveNext ()
System.Runtime.CompilerServices.AsyncVoidMethodBuilder:Start<WalletConnectSharp.Core.Controllers.TypedMessageHandler/<>c__DisplayClass18_0`2/<<HandleMessageType>g__InspectResponseRaw|2>d<WalletConnectSharp.Sign.Models.Engine.Methods.SessionPropose, WalletConnectSharp.Sign.Models.Engine.Methods.SessionProposeResponse>> (WalletConnectSharp.Core.Controllers.TypedMessageHandler/<>c__DisplayClass18_0`2/<<HandleMessageType>g__InspectResponseRaw|2>d<WalletConnectSharp.Sign.Models.Engine.Methods.SessionPropose, WalletConnectSharp.Sign.Models.Engine.Methods.SessionProposeResponse>&)
WalletConnectSharp.Core.Controllers.TypedMessageHandler/<>c__DisplayClass18_0`2<WalletConnectSharp.Sign.Models.Engine.Methods.SessionPropose, WalletConnectSharp.Sign.Models.Engine.Methods.SessionProposeResponse>:<HandleMessageType>g__InspectResponseRaw|2 (object,WalletConnectSharp.Core.Models.Relay.DecodedMessageEvent)
WalletConnectSharp.Core.Controllers.TypedMessageHandler/<RelayerMessageCallback>d__17:MoveNext ()
System.Runtime.CompilerServices.AsyncVoidMethodBuilder:Start<WalletConnectSharp.Core.Controllers.TypedMessageHandler/<RelayerMessageCallback>d__17> (WalletConnectSharp.Core.Controllers.TypedMessageHandler/<RelayerMessageCallback>d__17&)
WalletConnectSharp.Core.Controllers.TypedMessageHandler:RelayerMessageCallback (object,WalletConnectSharp.Core.Models.Relay.MessageEvent)
WalletConnectSharp.Core.Controllers.Relayer/<OnMessageEvent>d__80:MoveNext ()
System.Runtime.CompilerServices.AsyncTaskMethodBuilder:Start<WalletConnectSharp.Core.Controllers.Relayer/<OnMessageEvent>d__80> (WalletConnectSharp.Core.Controllers.Relayer/<OnMessageEvent>d__80&)
WalletConnectSharp.Core.Controllers.Relayer:OnMessageEvent (WalletConnectSharp.Core.Models.Relay.MessageEvent)
WalletConnectSharp.Core.Controllers.Relayer/<OnProviderPayload>d__77:MoveNext ()
System.Runtime.CompilerServices.AsyncTaskMethodBuilder:SetResult ()
WalletConnectSharp.Core.Controllers.Relayer/<AcknowledgePayload>d__81:MoveNext ()
System.Runtime.CompilerServices.AsyncTaskMethodBuilder:SetResult ()
WalletConnectUnity.Core.WebSocketConnection/<WalletConnectSharp-Network-IJsonRpcConnection-SendResult>d__38`1<bool>:MoveNext () (at ./Library/PackageCache/com.walletconnect.core@3.1.6/Runtime/WebSocketConnection.cs:129)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder:SetResult ()
NativeWebSocket.WebSocket/<SendMessage>d__35:MoveNext () (at ./Library/PackageCache/com.walletconnect.core@3.1.6/Runtime/External/NativeWebSocket/WebSocket.cs:604)
UnityEngine.UnitySynchronizationContext:ExecuteTasks () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:107)
```

```
Web3Exception: Error occured when tried to connect using WalletConnect.
ChainSafe.Gaming.WalletConnectUnity.WalletConnectUnityProvider.Connect () (at /Users/juanmanuelspoleti/Desktop/workspace/web3.unity/Packages/io.chainsafe.web3-unity/Runtime/WalletConnect/WalletConnectUnityProvider.cs:103)
ChainSafe.Gaming.Web3.Evm.Wallet.WalletSigner.WillStartAsync () (at <b5d0927d50444c5591da71b0458b3191>:0)
ChainSafe.Gaming.Web3.Web3.InitializeAsync () (at <b5d0927d50444c5591da71b0458b3191>:0)
ChainSafe.Gaming.Web3.Build.Web3Builder.LaunchAsync () (at <b5d0927d50444c5591da71b0458b3191>:0)
ChainSafe.Gaming.UnityPackage.Common.ILoginProvider.Login () (at /Users/juanmanuelspoleti/Desktop/workspace/web3.unity/Packages/io.chainsafe.web3-unity/Runtime/Common/Scripts/ILoginProvider.cs:52)
Scenes.LoginProvider.TryLogin () (at /Users/juanmanuelspoleti/Desktop/workspace/web3.unity/Packages/io.chainsafe.web3-unity/Runtime/Common/Scripts/LoginProvider.cs:49)
WalletConnectUnityLoginProvider.LoginClicked () (at Assets/Samples/web3.unity SDK/2.5.0/Web3.Unity Samples/Scripts/Scenes/WalletConnectUnityLoginProvider.cs:23)
System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__7_0 (System.Object state) (at <e307bbb467104258887a104f6151f183>:0)
UnityEngine.UnitySynchronizationContext+WorkRequest.Invoke () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:153)
UnityEngine.UnitySynchronizationContext.Exec () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:83)
UnityEngine.UnitySynchronizationContext.ExecuteTasks () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:107)
```
Screenshot 2024-07-01 at 3 53 54 PM

@oleksandrchainsafe it actually worked with Trust Wallet, but is not working with MetaMask that I believe is the most important one

**Comment by juans-chainsafe** [comment]
@oleksandrchainsafe I did some connections with the wallets that Wallet connect support, here is the report:

- Metamask ❌ 

- Trust wallet ✅ 
 - Binance Web3 wallet ✅ 
- Uniswap ❌ 
- Zerion ❌ 
- SafePal ✅ 
- Rainbow ❌ 
- Ledger Live ✅ 
- Coinbase Wallet ❔  (I couldn't connect coinbase wallet at all, but doesnt looks like our issue)

I dont count Coinbase Wallet, so we can connect to 4 wallets out of 8.
Success rate is 0.5

**Comment by kantagara** [comment]
@juans-chainsafe and what about the regular WC that is in 2.6?

**Comment by juans-chainsafe** [comment]
> @juans-chainsafe and what about the regular WC that is in 2.6?

Metamask ✅ 
Trust wallet ❌ 
Binance Web3 wallet ❌ 
Uniswap ❌ 
Zerion ❌ 
SafePal ✅ 
Rainbow ❌ 
Ledger Live ✅ 

3/8
So, the new implementation has 4/8, only 1 more, but doesnt connect with Metamask, which is the most important one

**Comment by juans-chainsafe** [comment]
> > @juans-chainsafe and what about the regular WC that is in 2.6?
> 
> Metamask ✅ Trust wallet ❌ Binance Web3 wallet ❌ Uniswap ❌ Zerion ❌ SafePal ✅ Rainbow ❌ Ledger Live ✅
> 
> 3/8 So, the new implementation has 4/8, only 1 more, but doesnt connect with Metamask, which is the most important one

I also want to add, that with this integration, takes more time to load the Wallet connect screen with the QR code, way more than before.

**Comment by oleksandrchainsafe** [comment]
It was decided we won't use this version for now, as it performs purely as described by Juan in previous messages. We are, however, going to keep this in the repo but disable it, so it's not accessible to the user.

**Comment by oleksandrchainsafe** [comment]
Closing PR but keeping the branch as a reference point. Might be useful if we decide to integrate it again some time in the future.

**Pull request closed without merging**